[testharness.js] `Test#add_cleanup` tests and fix

jugglinmike opened the pull request:
@jgraham The first commit is dedicated to improving coverage (in support of my upcoming work to recognize Promises returned from cleanup functions). The second is a behavior change--from the commit message body:

> Prior to application of this patch, `testharness.js` did not catch
> errors thrown by cleanup functions. In this case, script execution was
> interrupted, so additional cleanup functions were not invoked. Further,
> any subsequently-defined tests were not run, but neither were they
> detected or reported. The framework reported the error via the global
> `onerror` event, exposing the implementation-defined `message` property
> in the resulting failure report.
>
> Ensure that script evaluation proceeds following an error thrown by a
> cleanup function. This includes invocation of any remaining cleanup
> functions and detection of remaining tests. Additional tests are
> explicitly not executed due to the general instability of the global
> state under these conditions, but they are reported with the status
> "NOTRUN". Update the reported test harness error message for this
> situation to describe the situation in terms that are consistent across
> implementations.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/w3c/web-platform-tests/6151)
<!-- Reviewable:end -->

jugglinmike committed:
[testharness] Add tests for `Test#add_cleanup`

jugglinmike committed:
[testharness] Handle errors during cleanup

Prior to application of this patch, `testharness.js` did not catch
errors thrown by cleanup functions. In this case, script execution was
interrupted, so additional cleanup functions were not invoked. Further,
any subsequently-defined tests were not run, but neither were they
detected or reported. The framework reported the error via the global
`onerror` event, exposing the implementation-defined `message` property
in the resulting failure report.

Ensure that script evaluation proceeds following an error thrown by a
cleanup function. This includes invocation of any remaining cleanup
functions and detection of remaining tests. Additional tests are
explicitly not executed due to the general instability of the global
state under these conditions, but they are reported with the status
"NOTRUN". Update the reported test harness error message for this
situation to describe the situation in terms that are consistent across
implementations.

w3c-bots commented:
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/239659620)

# Firefox (nightly) #
Testing web-platform-tests at revision 578ae62f3febdb8f1cff115e4c04c3e686d9337b
Ignoring 4 changed files:
No files changed

w3c-bots commented:
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/239659621)

# Sauce (safari) #
Testing web-platform-tests at revision 578ae62f3febdb8f1cff115e4c04c3e686d9337b
Ignoring 4 changed files:
No files changed

w3c-bots commented:
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/239659627)

# Chrome (unstable) #
Testing web-platform-tests at revision 578ae62f3febdb8f1cff115e4c04c3e686d9337b
Ignoring 4 changed files:
No files changed

w3c-bots commented:
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/239659628)

# Sauce (MicrosoftEdge) #
Testing web-platform-tests at revision 578ae62f3febdb8f1cff115e4c04c3e686d9337b
Ignoring 4 changed files:
No files changed

gsnedders commented:
FWIW, I agreed the behaviour change was a good idea. I'm leaving the actual review to @jgraham tho given I've scarcely touched it in forever.

jugglinmike commented:
Thanks, @gsnedders!

jgraham commented:
Reviewed 4 of 4 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[resources/testharness.js, line 1735 at r1](https://reviewable.io:443/reviews/w3c/web-platform-tests/6151#-Km68feAZEVwo_-6-2YY:-Km68feBTTNL3uZxKPXc:b-35quf9) ([raw file](https://github.com/w3c/web-platform-tests/blob/636edb8d2055885c04b8e973616b61fe7994a9ff/resources/testharness.js#L1735)):*
> ```JavaScript
>             HAVE_TESTS:2,
>             HAVE_RESULTS:3,
>             ABORTED:4,
> ```

I would swap these unless there's some reason not to.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/w3c/web-platform-tests/6151)*
<!-- Sent from Reviewable.io -->

jugglinmike commented:
> I would swap these unless there's some reason not to.

Sure. Addressed with a `fixup!` commit.

jugglinmike commented:
Thanks, James! I've rebased this branch on the latest version of `master` and
squashed the commits as appropriate. The original version of the branch is
available here:

https://github.com/w3c/web-platform-tests/compare/master...bocoup:testharness-cleanup-tests-orig

If this looks good to you, would you mind merging on my behalf?
